File: resources/sdk/webmessagingjava/extensions/connector/okhttp/OkHttpClientConnector.java
package cloud.genesys.webmessaging.sdk.connector.okhttp;

import com.google.common.util.concurrent.SettableFuture;
import cloud.genesys.webmessaging.sdk.connector.ApiClientConnector;
import cloud.genesys.webmessaging.sdk.connector.ApiClientConnectorRequest;
import cloud.genesys.webmessaging.sdk.connector.ApiClientConnectorResponse;
import okhttp3.*;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.Map;
import java.util.concurrent.Future;

public class OkHttpClientConnector implements ApiClientConnector {
    private final OkHttpClient client;

    public OkHttpClientConnector(OkHttpClient client) {
        this.client = client;
    }

    @Override
    public ApiClientConnectorResponse invoke(ApiClientConnectorRequest request) throws IOException {
        Call call = client.newCall(buildRequest(request));
        return new OkHttpResponse(call.execute());
    }

    private Request buildRequest(ApiClientConnectorRequest request) throws IOException {
        Request.Builder builder = new Request.Builder()
                .url(request.getUrl());

        Map<String, String> headers = request.getHeaders();
        if (headers != null && !headers.isEmpty()) {
            for (Map.Entry<String, String> header : headers.entrySet()) {
                builder = builder.addHeader(header.getKey(), header.getValue());
            }
        }

        String method = request.getMethod();
        if ("GET".equals(method)) {
            builder = builder.get();
        }
        else if ("HEAD".equals(method)) {
            builder = builder.head();
        }
        else if ("POST".equals(method)) {
            builder = builder.post(createBody(request));
        }
        else if ("PUT".equals(method)) {
            builder = builder.put(createBody(request));
        }
        else if ("DELETE".equals(method)) {
            builder = builder.delete();
        }
        else if ("PATCH".equals(method)) {
            builder = builder.patch(createBody(request));
        }
        else {
            throw new IllegalStateException("Unknown method type " + method);
        }

        return builder.build();
    }

    private RequestBody createBody(ApiClientConnectorRequest request) throws IOException {
        String contentType = "application/json";
        Map<String, String> headers = request.getHeaders();
        if (headers != null && !headers.isEmpty()) {
            for (String name : headers.keySet()) {
                if (name.equalsIgnoreCase("content-type")) {
                    contentType = headers.get(name);
                    break;
                }
            }
        }
        if (request.hasBody()) {
            return RequestBody.create(MediaType.parse(contentType), request.readBody());
        }
        return RequestBody.create(null, new byte[0]);
    }

    @Override
    public void close() throws Exception { }
}

File: resources/sdk/purecloudkotlin/extensions/connector/okhttp/OkHttpClientConnector.kt
package com.mypurecloud.sdk.v2.connector.okhttp

import com.google.common.util.concurrent.SettableFuture
import com.mypurecloud.sdk.v2.AsyncApiCallback
import com.mypurecloud.sdk.v2.connector.ApiClientConnector
import com.mypurecloud.sdk.v2.connector.ApiClientConnectorRequest
import com.mypurecloud.sdk.v2.connector.ApiClientConnectorResponse
import com.squareup.okhttp.*
import java.io.IOException
import java.util.concurrent.Future

class OkHttpClientConnector(private val client: OkHttpClient) : ApiClientConnector {
    @Throws(IOException::class)
    override fun invoke(request: ApiClientConnectorRequest): ApiClientConnectorResponse {
        val call = client.newCall(buildRequest(request))
        return OkHttpResponse(call.execute())
    }

    override fun invokeAsync(request: ApiClientConnectorRequest, callback: AsyncApiCallback<ApiClientConnectorResponse?>): Future<ApiClientConnectorResponse> {
        val future = SettableFuture.create<ApiClientConnectorResponse>()
        try {
            val call = client.newCall(buildRequest(request))
            call.enqueue(object : Callback {
                override fun onFailure(request: Request, e: IOException) {
                    callback.onFailed(e)
                    future.setException(e)
                }

                @Throws(IOException::class)
                override fun onResponse(response: Response) {
                    val okHttpResponse = OkHttpResponse(response)
                    callback.onCompleted(okHttpResponse)
                    future.set(OkHttpResponse(response))
                }
            })
        } catch (exception: Throwable) {
            callback.onFailed(exception)
            future.setException(exception)
        }
        return future
    }

    @Throws(IOException::class)
    private fun buildRequest(request: ApiClientConnectorRequest): Request {
        var builder = Request.Builder()
                .url(request.url)
        val headers = request.headers
        if (headers.isNotEmpty()) {
            for ((key, value) in headers) {
                builder = builder.addHeader(key, value)
            }
        }
        val method = request.method
        builder = when (method) {
            "GET" -> {
                builder.get()
            }
            "HEAD" -> {
                builder.head()
            }
            "POST" -> {
                builder.post(createBody(request))
            }
            "PUT" -> {
                builder.put(createBody(request))
            }
            "DELETE" -> {
                builder.delete()
            }
            "PATCH" -> {
                builder.patch(createBody(request))
            }
            else -> {
                throw IllegalStateException("Unknown method type $method")
            }
        }
        return builder.build()
    }

    @Throws(IOException::class)
    private fun createBody(request: ApiClientConnectorRequest): RequestBody {
        var contentType = "application/json"
        val headers = request.headers
        if (headers.isNotEmpty()) {
            for (name in headers.keys) {
                if (name.equals("content-type", ignoreCase = true)) {
                    contentType = headers[name] ?: ""
                    break
                }
            }
        }
        return if (request.hasBody()) {
            RequestBody.create(MediaType.parse(contentType), request.readBody())
        } else RequestBody.create(null, ByteArray(0))
    }

    @Throws(Exception::class)
    override fun close() {
    }
}
